Framework/DataObjects/inc/MantidDataObjects/EventList.h

```diff
@@ -357,7 +327,37 @@ class DLLExport EventList : public Mantid::API::IEventList {
                           Mantid::Kernel::Unit *toUnit);
   void convertUnitsQuickly(const double &factor, const double &power);
 
+  HistogramData::Histogram histogram() const override;
+  HistogramData::Counts counts() const override;
+  HistogramData::CountVariances countVariances() const override;
+  HistogramData::CountStandardDeviations
+  countStandardDeviations() const override;
+  HistogramData::Frequencies frequencies() const override;
+  HistogramData::FrequencyVariances frequencyVariances() const override;
+  HistogramData::FrequencyStandardDeviations
+  frequencyStandardDeviations() const override;
+  const HistogramData::HistogramY &y() const override;
+  const HistogramData::HistogramE &e() const override;
+  HistogramData::HistogramY &mutableY() override;
+  HistogramData::HistogramE &mutableE() override;
+  Kernel::cow_ptr<HistogramData::HistogramY> sharedY() const override;
+  Kernel::cow_ptr<HistogramData::HistogramE> sharedE() const override;
+  void setSharedY(const Kernel::cow_ptr<HistogramData::HistogramY> &y) override;
+  void setSharedE(const Kernel::cow_ptr<HistogramData::HistogramE> &e) override;
+
+protected:
+  void checkHistogram(const HistogramData::Histogram &histogram) const override;
+  void checkWorksWithPoints() const override;
+
 private:
+  const HistogramData::Histogram &histogramRef() const override {
+    return m_histogram;
+  }
+  HistogramData::Histogram &mutableHistogramRef() override;
+
+  /// Histogram object holding the histogram data. Currently only X.
+  HistogramData::Histogram m_histogram;
```

Comment:
Yes. I don't like it, but with the current `MatrixWorkspace` class tree there is not really another option.

Note that `EventList::histogram()` returns a fully initialized `Histogram` (generated on the fly similar to `readY()`).

There is also the ugly interaction with the MRU, but we cannot get rid of that, unless we accept a significant increase in memory consumption by keeping all `Y` and `E` in `m_histogram` (generated and updated based on the list of events).